File: PJ/Java_version/src/timer/Timer.java
package timer;

import protocol.MyHost;

/**
 * 计时器。
 * 参考自https://blog.csdn.net/qq_33669680/article/details/78332899
 */
public class Timer extends Thread {

    private TimeModel timeModel;

    private MyHost myHost;

    public Timer(MyHost myHost, TimeModel timeModel) {
        this.myHost = myHost;
        this.timeModel = timeModel;
    }

    @Override
    public void run() {
        while (true) {
            int time = timeModel.getTime();
            if (time > 0) {
                try {
                    Thread.sleep(time * 1000);

                    System.out.println("\n");
                    if (myHost != null) {
                        System.out.println(myHost.getHostName() + "等待ACK超时");
                        myHost.timeOut();
                    }
//                    if (gbnHost != null) {
//                        System.out.println("GBN接收方等待ACK超时");
//                        gbnHost.timeOut(gbnHost.getDestPort());
//                    } else if (srHost != null) {
//                        System.out.println("SR接收方等待ACK超时");
//                        srHost.timeOut(srHost.getDestPort());
//                    }

                    timeModel.setTime(0);

                } catch (Exception e) {
                }
            }
        }
    }
}

File: PJ/GBN/timer/Timer.py
import threading
import time


class Timer(threading.Thread):
    def __init__(self, myHost, timeModel):
        super().__init__()
        self.myHost = myHost
        self.timeModel = timeModel

    def run(self):
        while True:
            TIME = self.timeModel.getTime()
            # print("TIME", TIME)
            if TIME > 0:
                try:
                    time.sleep(TIME)
                    # print("\n")
                    if self.myHost is not None:
                        print(self.myHost.getHostName(), "等待ACK超时")
                        self.myHost.timeOut()
                    self.timeModel.setTime(0)
                except Exception as e:
                    print(e)